recipes/z3/all/conanfile.py

```diff
@@ -66,64 +73,60 @@ def requirements(self):
-            "Visual Studio": "15.7",
-            "clang": "5",
-            "apple-clang": "10",
-        }
-
     def validate(self):
         if Version(self.version) >= "4.8.11":
-            if self.settings.compiler.get_safe("cppstd"):
-                tools.check_min_cppstd(self, "17")
+            if self.info.settings.compiler.cppstd:
+                check_min_cppstd(self, self._minimum_cpp_standard)
             compiler = self.settings.compiler
-            min_version = self._compilers_minimum_version\
-                .get(str(compiler), False)
-            if min_version:
-                if Version(compiler.version) < min_version:
-                    raise ConanInvalidConfiguration(
-                        f"{self.name} requires C++17, which {compiler} {compiler.version} does not support.")
+            minimum_version = self._compilers_minimum_version.get(str(self.info.settings.compiler), False)
+            if minimum_version:
+                if Version(self.info.settings.compiler.version) < minimum_version:
+                    raise ConanInvalidConfiguration(f"{self.ref} requires C++{self._minimum_cpp_standard}, which your compiler does not support.")
             else:
                 self.output.info(
                     f"{self.name} requires C++17. Your compiler is unknown. Assuming it supports C++17.")
+        if is_msvc(self) and self.info.options.shared:
+            raise ConanInvalidConfiguration(f"{self.ref} can not be built as shared on Visual Studio and msvc.")
+
+    def source(self):
+        get(self, **self.conan_data["sources"][self.version],
+                  destination=self.source_folder, strip_root=True)
+
+    def generate(self):
+        tc = CMakeToolchain(self)
+        tc.variables["PACKAGE_CUSTOM_DEFINITION"] = True
+        tc.variables["Z3_USE_LIB_GMP"] = self.options.multiprecision != "internal"
+        tc.variables["Z3_USE_LIB_MPIR"] = self.options.multiprecision == "mpir"
+        tc.variables["SINGLE_THREADED"] = not self.options.multithreaded
+        tc.variables["Z3_BUILD_LIBZ3_SHARED"] = self.options.shared
+        tc.variables["Z3_INCLUDE_GIT_HASH"] = False
+        tc.variables["Z3_INCLUDE_GIT_DESCRIBE"] = False
+        tc.variables["Z3_ENABLE_EXAMPLE_TARGETS"] = False
+        tc.variables["Z3_BUILD_DOCUMENTATION"] = False
+        if is_msvc(self):
+            tc.variables["USE_MSVC_RUNTIME_LIBRARY_DLL"] = not is_msvc_static_runtime(self)
+        tc.generate()
+        tc = CMakeDeps(self)
+        tc.generate()
+        tc = VirtualBuildEnv(self)
+        tc.generate(scope="build")
 
     def build(self):
         apply_conandata_patches(self)
 
         if self.options.multiprecision == "mpir":
-            tools.save(os.path.join(self._build_subfolder, "gmp.h"), textwrap.dedent("""\
+            save(self, os.path.join(self.build_folder, "gmp.h"), textwrap.dedent("""\
                 #pragma once
                 #include <mpir.h>
                 """))
```

Comment:
Any idea why this patch doesn't work ?

Reply:
take a look to patches which are parts of this trick

Reply:
Ok I see. I guess the  CONAN_PKG::mpi and CONAN_INCLUDE_DIRS are not valid anymore. Could you point me what should I replace this with ?

Reply:
How is it odd? Can I help in any way?

Reply:
Thanks @victorpaleologue ! I won't be able to go further with this PR but feel free to use my work if you want to go further with it!